### PR TITLE
[next] Relay GraphQL Batching

### PR DESCRIPTION
**Pull request by wyattjoh**
## What does this PR do?

Adds GraphQL batching support that is compatible with the client side relay transport: https://github.com/relay-tools/react-relay-network-layer/blob/master/src/express-middleware/graphqlBatchHTTPWrapper.js 

## How do I test this PR?

Submit a GraphQL query:

```json
[{"query": "{ me { id } }", "id": "1"}]
```

And see the response:

```json
[{"id": "1", "payload": {"data": {"me": null}}}]
```

### CODE REVIEW

**Comment by alinhle** [comment]
@wyattjoh The user information seems to be lost in this PR, can you take a look?

**Comment by wyattjoh** [comment]
Fixed @cvle!

File: src/core/server/app/middleware/graphqlBatch.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+function wrapResponse(req: Request, res: Response) {
+  // If the request is not an array, or has no elements, we should skip it.
+  if (!Array.isArray(req.body) || req.body.length === 0) {
+    return res;
+  }
+
+  // If the request is an array, but it does not have an ID field, then we
+  // should skip it.
+  const needsUpgrade = Boolean(typeof req.body[0].id !== "undefined");
+  if (!needsUpgrade) {
```
Comment (alinhle) [inline src/core/server/app/middleware/graphqlBatch.ts:15]: Upgrade usually means upgrading the connection to a websocket, maybe a more generic term like `skip` will do here too?

Comment (alinhle) [reply src/core/server/app/middleware/graphqlBatch.ts:15]: Not that this is really important xD

File: src/core/client/framework/lib/network/createNetwork.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import {
+  authMiddleware,
+  batchMiddleware,
+  cacheMiddleware,
+  RelayNetworkLayer,
+  retryMiddleware,
+  urlMiddleware,
+} from "react-relay-network-modern/es";
```
Comment (wyattjoh) [inline src/core/client/framework/lib/network/createNetwork.ts:8]: Why the `/es`?

File: src/core/client/stream/tabs/profile/queries/ProfileQuery.tsx [hunk 3]
```diff
@@ -64,6 +64,10 @@ const ProfileQuery: StatelessComponent<InnerProps> = ({
       assetID,
       assetURL,
     }}
+    cacheConfig={{
+      // TODO: enable caching after mutations are adapted.
+      force: true,
+    }}
```
Comment (wyattjoh) [inline src/core/client/stream/tabs/profile/queries/ProfileQuery.tsx:70]: What's this about?

File: src/core/server/app/middleware/graphqlBatch.ts (hunk 1 above)
Comment (wyattjoh) [reply src/core/server/app/middleware/graphqlBatch.ts:15]: "needsUpgrade" to a relay batch connection :P the term "upgrade" isn't reserved here.

File: src/types/react-relay-network-modern.d.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
+/* tslint:disable */
+
+declare module "react-relay-network-modern/es" {
```
Comment (wyattjoh) [inline src/types/react-relay-network-modern.d.ts:5]: Can we consider submitting these to @types? With a `TODO` to replace them in the future?

File: src/core/client/framework/lib/network/createNetwork.ts (hunk 2 above)
Comment (alinhle) [reply src/core/client/framework/lib/network/createNetwork.ts:8]: `es` means it contains es modules and have not been compiled down to  `commonjs` => allows tree shaking to do its work.

File: src/types/react-relay-network-modern.d.ts (hunk 4 above)
Comment (alinhle) [reply src/types/react-relay-network-modern.d.ts:5]: Of course

File: src/core/client/stream/tabs/profile/queries/ProfileQuery.tsx (hunk 3 above)
Comment (alinhle) [reply src/core/client/stream/tabs/profile/queries/ProfileQuery.tsx:70]: If we adapt the `CreateCommentMutation` to also insert into the my profile list, we can activate caching here. We should however first integrate pagination for the my comments list first, before we can adapt them appropriately.

**Review by alinhle — APPROVED** [review]

**Pull request merged**